### PR TITLE
[SYCL] Fix tests for assert

### PR DESCRIPTION
**Pull request by s-kanaev**
Fixes of regressions for intel/llvm#3767

### CODE REVIEW

**Comment by s-kanaev** [comment]
@AlexeySachkov , @Fznamznon , possibly some of the tests are for features which are owned by you, so please, pay extreme attention to changes of those tests in this patch.

**Comment by s-kanaev** [comment]
@intel/llvm-reviewers-runtime , ping

File: SYCL/lit.cfg.py [hunk 1]
```diff
@@ -281,6 +281,20 @@
 if find_executable('sycl-ls'):
     config.available_features.add('sycl-ls')
 
+llvm_spirv_path = find_executable("llvm-spirv")
+if llvm_spirv_path:
+  lit_config.note("Found llvm-spirv")
+  config.substitutions.append( ('%LLVM_SPIRV', os.path.realpath(llvm_spirv_path)) )
+else:
+  lit_config.warning("Can't find llvm_spirv")
```
Comment (dm-vodopyanov) [inline SYCL/lit.cfg.py]:
```suggestion
  lit_config.warning("Can't find llvm-spirv")
```

Comment (s-kanaev) [reply SYCL/lit.cfg.py]: Fixed.

File: SYCL/lit.cfg.py [hunk 2]
```diff
@@ -281,6 +281,20 @@
 if find_executable('sycl-ls'):
     config.available_features.add('sycl-ls')
 
+llvm_spirv_path = find_executable("llvm-spirv")
```
Comment (vladimirlaz) [inline SYCL/lit.cfg.py]: why these checks are added here. As far AFAIR these tools are always present in the compiler. So the check is redundant.

In any case, I do not see any usages of the substitutions defined here

Comment (s-kanaev) [reply SYCL/lit.cfg.py]: `llvm-spirv` and `llvm-link` are used in SYCL/AOT/multiple-devices.cpp

Comment (vladimirlaz) [reply SYCL/lit.cfg.py]: The tools do but substitutions that have been defined don't. Also if it is expected that the tools are present conditionally it is reasonable to add REQUIRES: lvm-spirv, etc to corresponding tests

Comment (s-kanaev) [reply SYCL/lit.cfg.py]: > but substitutions that have been defined don't

It's explicitly `%LLVM_LINK` and `%LLVM_SPIRV` which are used now in `multiple-devices.cpp`

> REQUIRES: lvm-spirv

is there already a feature defined as `llvm-spirv` or `llvm-link`? Is there any simple example of defining such a feature?

Comment (vladimirlaz) [reply SYCL/lit.cfg.py]: you can use sycl-ls as example

**Review by vzakhari — DISMISSED** [review]
DeviceLib change LGTM.

File: SYCL/AOT/multiple-devices.cpp [hunk 3]
```diff
@@ -18,11 +18,19 @@
 // RUN: %GPU_RUN_PLACEHOLDER %t_all.out
 // RUN: %ACC_RUN_PLACEHOLDER %t_all.out
 
+// FIXME: Change the behaviour when proper automaton for assert support is
```
Comment (vladimirlaz) [inline SYCL/AOT/multiple-devices.cpp]:
```suggestion
// FIXME: Change the behavior when proper automation for assert support is
```

File: SYCL/lit.cfg.py [hunk 4]
```diff
@@ -281,6 +281,20 @@
...
+if llvm_spirv_path:
+  lit_config.note("Found llvm-spirv")
+  config.substitutions.append( ('%LLVM_SPIRV', os.path.realpath(llvm_spirv_path)) )
```
Comment (vladimirlaz) [inline SYCL/lit.cfg.py]: please use format similar to other substitutions (small letters)

File: SYCL/lit.cfg.py (hunk 2 above)
Comment (vladimirlaz) [reply SYCL/lit.cfg.py]: feature-related comment still not applied

**Review by vladimirlaz — CHANGES_REQUESTED** [review]

File: SYCL/AOT/multiple-devices.cpp (hunk 3 above)
Comment (s-kanaev) [reply SYCL/AOT/multiple-devices.cpp]: Fixed.

File: SYCL/lit.cfg.py (hunk 2 above)
Comment (s-kanaev) [reply SYCL/lit.cfg.py]: Fixed.

File: SYCL/lit.cfg.py (hunk 4 above)
Comment (s-kanaev) [reply SYCL/lit.cfg.py]: Fixed.

**Review by vladimirlaz — DISMISSED** [review]

**Review by romanovvlad — COMMENTED** [review]
LGTM

**Review by AGindinson — DISMISSED** [review]

**Review by dm-vodopyanov — DISMISSED** [review]

File: SYCL/AOT/multiple-devices.cpp [hunk 5]
```diff
@@ -18,11 +18,19 @@
 // RUN: %GPU_RUN_PLACEHOLDER %t_all.out
 // RUN: %ACC_RUN_PLACEHOLDER %t_all.out
 
+// FIXME: Change the behavior when proper automaton for assert support is
```
Comment (mlychkov) [inline SYCL/AOT/multiple-devices.cpp]:
```suggestion
// FIXME: Change the behavior when proper automation for assert support is
```

File: SYCL/lit.cfg.py [hunk 6]
```diff
@@ -281,6 +281,22 @@
...
+else:
+  lit_config.warning("Can't find llvm-spirv")
+
+llvm_link_path = find_executable("llvm-link")
+if llvm_link_path:
+  lit_config.note("Found llvm-link")
+  config.available_features.add('llvm-link')
+  config.substitutions.append( ('%llvm_link', os.path.realpath(llvm_link_path)) )
+else:
+  lit_config.warning("Can't find llvm-link")
```
Comment (mlychkov) [inline SYCL/lit.cfg.py]:
```suggestion
llvm_tools = ["llvm-spirv", "llvm-link"]
for llvm_tool in llvm_tools:
  llvm_tool_path = find_executable(llvm_tool)
  if llvm_tool_path:
    lit_config.note("Found " + llvm_tool)
    config.available_features.add(llvm_tool)
    config.substitutions.append( ('%' + llvm_tool.replace('-', '_'),
                                  os.path.realpath(llvm_tool_path)) )
  else:
    lit_config.warning("Can't find " + llvm_tool)
```

**Review by mlychkov — DISMISSED** [review]
Just minor improvements. Other LGTM.

**Review by MrSidims — DISMISSED** [review]
overall LGTM

**Review by kbobrovs — DISMISSED** [review]
ESIMD test changes LGTM

**Review by mlychkov — APPROVED** [review]

**Review by dm-vodopyanov — APPROVED** [review]

**Review by MrSidims — APPROVED** [review]

**Review by vladimirlaz — APPROVED** [review]

**Review by AGindinson — APPROVED** [review]

File: SYCL/AOT/multiple-devices.cpp [hunk 7]
```diff
@@ -18,11 +18,19 @@
 // RUN: %GPU_RUN_PLACEHOLDER %t_all.out
 // RUN: %ACC_RUN_PLACEHOLDER %t_all.out
 
+// FIXME: Change the behavior when proper automation for assert support is
+// introduced
```
Comment (Fznamznon) [inline SYCL/AOT/multiple-devices.cpp:22]: Will it be fixed by https://github.com/intel/llvm/pull/3767 ?

File: SYCL/Basic/device_code_dae.cpp [hunk 8]
```diff
@@ -2,8 +2,10 @@
 // driver
 // UNSUPPORTED: cuda
 // CUDA does not support SPIR-V.
-// RUN: %clangxx -fsycl-device-only -Xclang -fenable-sycl-dae -Xclang -fsycl-int-header=int_header.h %s -c -o device_code.bc -Wno-sycl-strict
-// RUN: %clangxx %cxx_std_optionc++17 %include_option int_header.h %debug_option -c %s -o host_code.o %sycl_options -Wno-sycl-strict
+// FIXME Disabled fallback assert as it'll require either online linking or
+// explicit offline linking step here
```
Comment (Fznamznon) [inline SYCL/Basic/device_code_dae.cpp:6]: Why not to add explicit offline linking like in previous test?

File: SYCL/Basic/fpga_tests/fpga_aocx.cpp [hunk 9]
```diff
@@ -11,15 +11,17 @@
 /// E2E test for AOCX creation/use/run for FPGA
 // Produce an archive with device (AOCX) image. To avoid appending objects to
 // leftover archives, remove one if exists.
+// FIXME Disabled use of devicelib by assert feature until the 2-step build gets
+// fixed.
```
Comment (Fznamznon) [inline SYCL/Basic/fpga_tests/fpga_aocx.cpp:15]: Will it be fixed by https://github.com/intel/llvm/pull/3767 ?

Comment (s-kanaev) [reply SYCL/Basic/fpga_tests/fpga_aocx.cpp:15]: This one won't be fixed by intel/llvm#3767. There's an issue in driver which isn't connected to intel/llvm#3767. The issue can be encountered when using any devicelib function and compiling the same way it's done in this test. @mdtoguchi  is aware of the issue.

File: SYCL/Basic/device_code_dae.cpp (hunk 8 above)
Comment (s-kanaev) [reply SYCL/Basic/device_code_dae.cpp:6]: Per offline discussion, agreed to add offline linking to this test in distinct patch.

File: SYCL/AOT/multiple-devices.cpp (hunk 7 above)
Comment (s-kanaev) [reply SYCL/AOT/multiple-devices.cpp:22]: Per offline discussion, agreed to elaborate on the comment in distinct patch.

**Review by Fznamznon — APPROVED** [review]

**Review by AlexeySachkov — APPROVED** [review]

**Pull request merged**